heron/scheduler-core/src/java/com/twitter/heron/scheduler/SubmitterMain.java

```diff
@@ -427,42 +428,37 @@ public boolean submitTopology() {
       SysUtils.closeIgnoringExceptions(launcher);
       SysUtils.closeIgnoringExceptions(statemgr);
     }
-
-    return isSuccessful;
   }
 
-  protected boolean validateSubmit(SchedulerStateManagerAdaptor adaptor, String topologyName) {
+  protected void validateSubmit(SchedulerStateManagerAdaptor adaptor, String topologyName)
+      throws TopologySubmissionException {
     // Check whether the topology has already been running
     Boolean isTopologyRunning = adaptor.isTopologyRunning(topologyName);
 
     if (isTopologyRunning != null && isTopologyRunning.equals(Boolean.TRUE)) {
-      LOG.severe("Topology already exists");
-      return false;
+      throw new TopologySubmissionException(
+          String.format("Topology '%s' already exists", topologyName));
     }
-
-    return true;
   }
 
-  protected URI uploadPackage(IUploader uploader) {
+  protected URI uploadPackage(IUploader uploader) throws UploaderException {
     // initialize the uploader
     uploader.initialize(config);
 
     // upload the topology package to the storage
-    URI uploaderRet = uploader.uploadPackage();
-
-    return uploaderRet;
+    return uploader.uploadPackage();
   }
 
-  protected boolean callLauncherRunner(Config runtime) {
+  protected void callLauncherRunner(Config runtime, String topologyName)
+      throws TopologySubmissionException {
     // using launch runner, launch the topology
     LaunchRunner launchRunner = new LaunchRunner(config, runtime);
     boolean result = launchRunner.call();
 
     // if failed, undo the uploaded package
     if (!result) {
-      LOG.severe("Failed to launch topology. Attempting to roll back upload.");
-      return false;
+      throw new TopologySubmissionException(
+          String.format("Failed to launch topology '%s'", topologyName));
```

Comment:
This is fixed. I refactored down to [`LaunchRunner.java`](https://github.com/twitter/heron/pull/1593/files#diff-bc651b82b5edf290148aedd84d4b8633) in #1593. However, notice that I did not not go even deeper down to `setPackingPlan`, `setExecutionState`, etc because this anti-pattern is deeply nested. Not sure if we can stop at the newly `if-null-then-throw-exception` pattern 😅